---
title: "ETO4 Preliminary RNA seq"
author: 
  - Noah Siegel
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Learning objectives:

* Create a gene-level count matrix of Salmon quantification using tximport
* Perform differential expression of a single factor experiment in DESeq2
* Perform quality control and exploratory visualization of RNA-seq data in R


## Tximeta

We first need to read our data into R. To do that, we will use a package called
[tximeta](https://bioconductor.org/packages/release/bioc/html/tximeta.html).
We use tximeta for two main reasons: first, it facilitates summarizing transcript-level counts
from Salmon to the gene-level for differential expression analysis. Second, tximeta enhances
incorporates metadata (e.g. transcript locations, transcript and genome source and version, 
appropriate chromosome lengths, etc) for each transcriptome. This ensures computational reproducibility 
by attaching critical annotation information to the data object, such that exact quantifications can be 
reproduced from raw data (all software versions are also attached to the data object).

For more information, see the
[tximeta vignette](https://bioconductor.org/packages/release/bioc/vignettes/tximeta/inst/doc/tximeta.html)

Since we're working in binder, all of the software is already installed. If you'd
like to do a similar analysis on your own system, you can use the following 
installation commands:

```{r install, eval = F}
if (!requireNamespace("BiocManager", quietly = TRUE))
 install.packages("BiocManager")
BiocManager::install("tximeta", version = "3.10")
BiocManager::install("DESeq2")
install.packages("ggplot2")
install.packages("dplyr")
install.packages("readr")
install.packages("pheatmap")
```

```{r library, message=FALSE}
library(SummarizedExperiment)
library(tximeta)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(readr)
library(pheatmap)
```

```{r, echo = F, message = F}
library(kableExtra)
```

```{r}
samples <- read_csv("rnaseq_samples.csv")
```
List samples

```{r, echo = F}
kableExtra::kable(samples) %>%
  kable_styling()
```


Define reference files

```{r}
indexDir <- file.path("rnaseq", "quant", "sc_ensembl_index")
fastaFTP <- c("ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz")
gtfPath <- "ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz"
gtfLocal <- "human_ref/Homo_sapiens.GRCh38.99.gtf.gz"
fastaLocal <- "human_ref/Homo_sapiens.GRCh38.cdna.all.fa.gz"
```

Build object
```{r, warning = F}
makeLinkedTxome(indexDir = indexDir,
                source   = "Ensembl", 
                organism = "Homo sapiens", 
                release  = "99", 
                fasta    = fastaLocal,
                gtf      = gtfLocal,
                genome   = "GCF_000001405.39",
                write    = FALSE)
```

To create an `se` object for using our `tximeta` information:
```{r, warning = F}
se <- tximeta(samples)
```

To look at our `se` object, we can use the following commands:
```{r, eval = F, eval = F}
colData(se)
assayNames(se)
rowRanges(se)
seqinfo(se)
```

And to summarize to gene level, we can use `summarizeToGene()`.

```{r stg}
gse <- summarizeToGene(se)
```

Data

```{r, eval = F}
rowRanges(gse)
mcols(gse)
```

## Differential Expression with `DESeq2`

Create DESeq object
```{r dds_from_set}
dds <- DESeqDataSet(se = gse, design = ~condition)
```

Now that we have a DESeq2 object, we can can perform differential expression.

```{r deseq}
dds <- DESeq(dds)
```

Everything from normalization to linear modeling was carried out by the use of a single function! This function prints out a message for the various steps it performs.

And look at the results! The results() function lets you extract the base means across samples, log2-fold changes, standard errors, test statistics etc. for every gene.

```{r, warning = F}
res <- results(dds)
```

```{r, eval = F}
head(res)
```

```{r, echo= F}
kableExtra::kable(head(res)) %>%
  kable_styling()
```

The first thing that prints to the screen is information about the "contrasts"
in the differential expression experiment. By default, DESeq2 selects the 
alphabetically first factor to the be "reference" factor. Here that doesn't 
make that much of a difference. However, it does change how we interpret the
log2foldchange values. We can read these results as, "Compared to *SNF2* mutant,
WT had a decrease of -0.2124 in log2fold change of gene expression. 

Speaking of log2fold change, what do all of these columns mean?

| baseMean   | giving means across all samples |
|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| log2FoldChange | log2 fold changes of gene expression from one condition to another. Reflects how different the expression of a gene in one condition is from the expression of the same gene in another condition. |
| lfcSE  | standard errors  (used to calculate p value) |
| stat | test statistics used to calculate p value) |
| pvalue | p-values for the log fold change |
| padj | adjusted p-values |


Filter by adjusted p value or log2 Fold Change:

```{r}
res_sig <- subset(res, padj<.05)
res_lfc <- subset(res_sig, abs(log2FoldChange) > 1) 
```

```{r, eval = F}
head(res_lfc)
```

```{r, echo= F}
kableExtra::kable(head(res_lfc))  %>%
  kable_styling()
```

## Visualization of RNA-seq and Differential Expression Results

Looking at our results is great, but visualizing them is even better!

### MA Plot

The MA plot provides a global view of the relationship between the expression change between conditions (log ratios, M), the average expression strength of the genes (average mean, A) and the ability of the algorithm to detect differential gene expression: genes that pass the significance threshold are colored in red

```{r ma_plt}
plotMA(res)
```
 

### Plotting individual genes

Il22Ra
```{r gene_count}
# Il22Ra expression
plotCounts(dds, gene='ENSG00000142677', intgroup="condition")
```

### Normalization & Transformation

`DESeq2` automatically normalizes our count data when it runs differential 
expression. However, for certain plots, we need to normalize our raw count data.
One way to do that is to use the `vst()` function. It perform variance 
stabilized transformation on the count data, while controlling for library
size of samples.

```{r vsd}
vsd <- vst(dds)
```

### MDS Plot

An MDS (multi-dimensional scaling) plot gives us an idea of how our samples relate to each other. The closer two 
samples are on a plot, the more similar all of their counts are. To generate
this plot in DESeq2, we need to calculate "sample distances" and then plot them.

```{r mds_filt}
# calculate sample distances
sample_dists <- assay(vsd) %>%
  t() %>%
  dist() %>%
  as.matrix() 
```

```{r, eval = F}
head(sample_dists)
```

```{r, echo = F}
kableExtra::kable(head(sample_dists))  %>%
  kable_styling()
```

Next, let's calculate the MDS values from the distance matrix.
```{r}
mdsData <- data.frame(cmdscale(sample_dists))
mds <- cbind(mdsData, as.data.frame(colData(vsd))) # combine with sample data
```

```{r, eval = F}
head(mds)
```

```{r, echo = F}
kableExtra::kable(head(mds))%>%
  kable_styling()
```

And plot with `ggplot2`!

```{r}
ggplot(mds, aes(X1, X2, shape = condition)) + 
  geom_point(size = 3) +
  theme_minimal()
``` 


**Question**  
How similar are the samples between conditions?

### Heatmap

Top 20 genes with the largest postivie log2fold change

```{r}
genes <- order(res_lfc$log2FoldChange, decreasing=TRUE)[1:20]
```

We can also make a data.frame that contains information about our samples that 
will appear in the heatmap. We will use our samples data.frame from before to
do this.

```{r}
annot_col <- samples %>%
        tibble::column_to_rownames('names') %>%
        select(condition) %>%
        as.data.frame()
```

```{r, eval = F}
head(annot_col)
```

```{r, echo = F}
kableExtra::kable(head(annot_col)) %>%
  kable_styling()
```

And now plot the heatmap!

```{r}
pheatmap(assay(vsd)[genes, ], cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=FALSE, annotation_col=annot_col)
```
